regression/ansi-c/sizeof6/main.c

```diff
@@ -0,0 +1,23 @@
+int main()
+{
+  long long i;
+#ifndef _MSC_VER
+  _Static_assert(sizeof(int) == sizeof(*(1 ? ((void *)(0ll)) : (int *)1)), "");
+  // We are able to simplify all of the expressions involving i below to 0, but
+  // GCC and Clang don't do so. Hence, the static asserts pass for those
+  // compilers.
+  _Static_assert(
+    sizeof(int) != sizeof(*(1 ? ((void *)(i * 0)) : (int *)1)), "");
+  _Static_assert(
+    sizeof(int) != sizeof(*(1 ? ((void *)(i - i)) : (int *)1)), "");
+  _Static_assert(
+    sizeof(int) != sizeof(*(1 ? ((void *)(i ? 0ll : 0ll)) : (int *)1)), "");
+  _Static_assert(
+    sizeof(int) != sizeof(*(1 ? ((void *)(0 ? i : 0ll)) : (int *)1)), "");
```

Comment:
These appear to assert that some part of the subexpression is a compile time constant. May I suggest to use `__builtin_constant_p` for that. The sizeof assertion is a distraction.

Reply:
This, I'm afraid, just uncovered another challenge: the above "true" cases aren't deemed null-pointer constants by the compilers (and, therefore, the type of the `?:` is `void*` rather than `int*`). `__builtin_constant_p`, however, does return `1` for the very same expressions!

Reply:
I have added a commit that better exercises (and extends) an existing test to also document (and test) these differences.

Reply:
Uhhh, sad

Reply:
Might be worth filing a bug report with those compilers.